src/nodes.coffee

```diff
@@ -1152,25 +1192,75 @@ exports.Class = class Class extends Base
     args  = []
     o.classScope = func.makeScope o.scope
 
+    superClass = null
+    if @parent
+      superClass = o.classScope.freeVariable 'superClass', reserve: no
+      func.params.push new Param new IdentifierLiteral superClass
+      args.push @parent
+
     @hoistDirectivePrologue()
     @setContext name
     @walkBody name, o
-    @ensureConstructor name
+    @ensureConstructor()
     @addBoundFunctions o
+    @ensureConstructorSuperCall() if @parent
+
+    esClass = new EsClass name, superClass, @ctor
+    @hoistConstructorComment esClass
+
     @body.spaced = yes
+    @body.expressions.unshift @directives..., esClass.placeholder
     @body.expressions.push lname
+    @body.expressions.splice @body.expressions.indexOf(@ctor), 1, esClass
 
-    if @parent
-      superClass = new IdentifierLiteral o.classScope.freeVariable 'superClass', reserve: no
-      @body.expressions.unshift new Extends lname, superClass
-      func.params.push new Param superClass
-      args.push @parent
+    klass  = new Parens new Call func, args
+    klass  = new Assign @variable, klass, null, { @moduleDeclaration } if @variable
+    result = klass.compileToFragments o
+    esClass.placeholder.updateFragments result
+    result
+
+#### ES2015+ Classes
+
+# An EsClass represents an ECMAScript class definition.
+# Initialize an **EsClass** with its name, an optional superclass,
+# and an optional constructor.
+exports.EsClass = class EsClass extends Base
+  class Placeholder extends Base
+    constructor: ->
+      @nonce     = {}
+      @fragments = []
+
+    isStatement: YES
+
+    compileToFragments: (o) ->
+      @nonce
+
+    updateFragments: (fragments) ->
+      fragments.splice fragments.indexOf(@nonce), 1, @fragments...
+
+  children: ['ctor', 'comment']
 
-    @body.expressions.unshift @directives...
+  constructor: (@name, @parent, @ctor, @comment) ->
+    @placeholder = new Placeholder
 
-    klass = new Parens new Call func, args
-    klass = new Assign @variable, klass, null, { @moduleDeclaration } if @variable
-    klass.compileToFragments o
+  isStatement: YES
+
+  compileNode: (o) ->
+    code = @placeholder.fragments
+
+    code.push @makeCode "#{@tab}class #{@name} "
+    code.push @makeCode "extends #{@parent} " if @parent
+    code.push @makeCode '{'
+
+    unless @ctor.params.length is 0 and @ctor.body.isEmpty() and not @comment
+      o.indent += TAB
+      code.push @comment.compileToFragments(o)... if @comment
+      code.push @makeCode('\n'), @ctor.compileToFragments(o)..., @makeCode "\n#{@tab}"
+
+    code.push @makeCode '}'
+    code.push @makeCode "\n#{@tab}#{@name}.__super__ = #{@parent}.prototype;" if @parent
+
+    []
```

Comment:
It is! This is part of the 'hack' to hoist the constructor, whilst also ensuring it has the correct scope. Consider:

```coffee
class A
  @a = 1
```

In CS1.x this becomes:

```js
A = (function () {
  A.a = 1

  function A () {}

  return A;
})()
```

This works fine due to named function hoisting. Unfortunately `class`es are not hoisted, so the same approach won't work. We also can't just move the node to the top, because the scope must correspond to the constructors position in the body, e.g.:

```coffee
class A
  @a = 1
  a = 1
  constructor: -> a = 2
```

The `a` in the constructor should reference the outer scope, but if we move the `class` above the `@a = 1` then the constructor will have the wrong scope.

I couldn't find a mechanism in the compiler for dealing with this, so I resorted to compiling a 'placeholder' node in the target position, and leaving the constructor node in the source position. Once compilation is done, the placeholder is swapped out for the compiled by the constructor.

In this function, rather than returning the fragments, they are being attached to the placeholder for later substitution, and so for this location no fragments are returned.

Regrettably this still leaves a blank line in the source where the constructor was. Now I think about it, the same approach could be used here: return a 'special' fragment that can be spliced out later.

If you have any suggestions for dealing with this (namely, compiling a node with a scope that's not available until further down the tree), or ways of clarifying what's here, I'd gladly hear them.